MREUnityRuntime/MREUnityRuntimeLib/Core/Physics/JitterBuffer.cs

```diff
@@ -232,235 +320,577 @@ public void add(float bufferTime, float meanBufferTime, float targetBufferTime,
+			if (_snapshots.Count == 0)
+			{
+				HasUpdate = false;
+				CurrentLocalTime = nextTimestamp;
+
+				return;
+			}
+
+			float appliedSnapshotTimestamp = float.MinValue;
+			bool reset = false;
+			int index = 0;
+
+			while (index < _snapshots.Count && _snapshots.Keys[index] - nextTimestamp <= _timePrecision)
+			{
+				updateRigidBodies(new SnapshotSource(_snapshots.Values[index]));
+
+				reset = reset || _snapshots.Values[index].Flags == Snapshot.SnapshotFlags.ResetJitterBuffer;
+				appliedSnapshotTimestamp = _snapshots.Keys[index];
+				index++;
+			}
+
+			if (Math.Abs(appliedSnapshotTimestamp - nextTimestamp) <= _timePrecision)
+			{
+				// we have matching snapshot
+				CurrentLocalTime = appliedSnapshotTimestamp;
+			}
+			else
+			{
+				CurrentLocalTime = nextTimestamp;
+
+				if (appliedSnapshotTimestamp < nextTimestamp)
+				{
+					// if there are more snapshots so we can interpolate
+					if (index != 0 && index < _snapshots.Count && _snapshots.Count > 1)
+					{
+						updateRigidBodies(new InterpolationSource(_snapshots.Values[index-1], _snapshots.Values[index], nextTimestamp));
 					}
 				}
 			}
 
-			/// <summary>
-			/// Source identifier.
-			/// </summary>
-			public Guid Id { get; private set; }
+			// delete processed snapshots
+			for (int r = 0; r < index; r++) _snapshots.RemoveAt(0);
 
-			/// <summary>
-			/// This source snapshots.
-			/// </summary>
-			private SnapshotBuffer SnapshotBuffer = new SnapshotBuffer();
+			if (reset)
+			{
+				reset = false;
+				_runningStats.Reset();
+				_areAllRigidBodiesSleeping = true;
+			}
+		}
 
-			/// <todo>
-			/// can this be avoided?
-			/// </todo>
-			private Mode _mode = Mode.Init;
+#endregion
 
-			/// <summary>
-			/// Calc running average buffered time and deviation.
-			/// </summary>
-			private RunningStats _stats = new RunningStats();
+#region Iterator
 
-			/// <summary>
-			/// todo: add comment
-			/// </summary>
-			private float _speedUpCoef = 0.2f;
+		int _iteratorIndex = 0;
 
-			/// <summary>
-			/// todo: add comment
-			/// </summary>
-			private float _speedDownCoef = 0.5f;
+		internal void ResetIterator()
+		{
+			_iteratorIndex = 0;
+		}
 
-			/// <summary>
-			/// Specifies if there is an update since the last step.
-			/// </summary>
-			public bool HasUpdate = false;
+		internal bool Find(Guid key, out RigidBodyData rigidBodyData)
+		{
+			while (_iteratorIndex < _rigidBodies.Count && _rigidBodies.Keys[_iteratorIndex].CompareTo(key) < 0) _iteratorIndex++;
 
-			/// <summary>
-			/// Timestamp of the snapshot in local time of the source.
-			/// </summary>
-			float CurrentLocalTime = float.MinValue;
+			if (_iteratorIndex < _rigidBodies.Count &&
+				_rigidBodies.Keys[_iteratorIndex] == key &&
+				_rigidBodies.Values[_iteratorIndex].Time >= 0)
+			{
+				rigidBodyData = _rigidBodies.Values[_iteratorIndex++];
+				return true;
+			}
+			else
+			{
+				rigidBodyData = null;
+				return false;
+			}
```

Comment:
This whole thing can be simplified with LINQ for clarity.

`var rbPair = _rigidBodies.Where(rb => rb.Key == key).FirstOrDefault();
   if (rbPair.Key == key && rbPair.Value.Time >= 0)
   {
    rigidBodyData = rbPair.Value;
    return true;
   }

   rigidBodyData = null;
   return false;`